TNL-715: Exit focus from annotated text

jmclaus opened the pull request:
**Ticket:** [TNL-715](https://openedx.atlassian.net/browse/TNL-715)

Allows quick exiting from highlighted text tabbing by allowing to focus on the Show/Hide notes via the CTRL+OPTION+n keyboard shortcut.

@polesye @cptvitamin Please review.

polesye commented on lms/static/js/edxnotes/plugins/accessibility.js:
```diff
@@ -148,6 +148,12 @@ define(['jquery', 'underscore', 'annotator'], function ($, _, Annotator) {
                     this.annotator.viewer.hide();
                     break;
+                // Gives focus to the show/hide notes button
+                case KEY.n:
```
I think we should bind this code directly to `document`.

polesye commented:
@jmclaus please rebase this branch.

jmclaus commented:
@polesye Done.

cptvitamin commented:
@jmclaus, Can we move focus to the Main Content area after toggling the notes on or off?  Since one can only annotate the main content area, it seems like a logical thing to do.  Currently, when a keyboard or screen reader user  toggles them on or off, focus remains on that button.  Then they have to manually find their way back to the main content.  This could make the page jump depending on the length of the content, which may not be desireable.  Not sure if you are tracking keyboard focus the entire time one is on the page, but if so, ideally, you would move focus to that location (which shouldn't result in a page jump).  If all of this seems like too much trouble, I'm willing to let it go.  There are mechanisms for screen reader users to move to main content area, and its not too inconvenient for keyboard users who could find their way to the Skip to main content link.

Currently, the "Save" button on the note window is a link.  Could we make that a button?  Could you also add SR text (```Save <span class="sr">Note</span>```) so that the screen reader says Save Note and not just Save which could be ambiguous if they use form field navigation to move focus to the button?

polesye commented on lms/static/js/edxnotes/plugins/accessibility.js:
```diff
@@ -148,6 +148,12 @@ define(['jquery', 'underscore', 'annotator'], function ($, _, Annotator) {
+                // Gives focus to the show/hide notes button
+                case KEY.n:
+                    if (event.ctrlKey && event.altKey)  {
+                        $('.action-toggle-notes').focus();
```
What if we will simply `disable/enable` notes on `Ctrl + Option/Alt + n` without moving the focus to `.action-toggle-notes` element?
@cptvitamin what do you think?

polesye replied:
We already use [aria-live="polite"](https://github.com/edx/edx-platform/blob/jmclaus/exit-focus-highlighted-text/lms/templates/edxnotes/toggle_notes.html#L11), so the users will be informed about what is happening.

cptvitamin replied:
@polesye I love the idea of a dedicated shortcut for toggling notes (we should talk about different ways for informing all users that this shortcut exits.)  However I do worry a bit about choosing a key combo that doesn't collide with any native functions of either the OS, browser or screen reader.   A quick search didn't turn up anything, so I say let's give it a try!

polesye commented:
> Currently, the "Save" button on the note window is a link. Could we make that a button? Could you also add SR text (Save <span class="sr">Note</span>) so that the screen reader says Save Note and not just Save which could be ambiguous if they use form field navigation to move focus to the button?

Done as a part of #6778 . See  https://github.com/edx/edx-platform/commit/d6a1e8b3140aa1b5a56d760ee3e1d31a9d9500c4

jmclaus commented:
@cptvitamin 
> I love the idea of a dedicated shortcut for toggling notes (we should talk about different ways for informing all users that this shortcut exits.) However I do worry a bit about choosing a key combo that doesn't collide with any native functions of either the OS, browser or screen reader. A quick search didn't turn up anything, so I say let's give it a try!

Indeed, a nice idea @polesye. It's done, when the `toggle notes` button is visible, pressing `CTRL+OPTION+n`will show/hide the notes directly without focusing on the button first.

jmclaus commented:
@cptvitamin We added a sandbox to test that feature:

http://sna11y.m.sandbox.edx.org/courses/edX/DemoX/Demo_Course/

cptvitamin commented:
@polesye @jmclaus, I have some bad news on the CTRL+ALT+N idea.  it works great for keyboard only users, however on Mac, CTRL+ALT is called the VoiceOver key and is used for basically all screen reader commands.  There is no CTRL+ALT+N command in Voiceover, which is why I thought this would work, but when I tested it out, it turns out that the CTRL+ALT never makes it to the browser.  It is intercepted by the screenreader/OS.  So the keyboard shortcut did not work...

Here are some other notes @clrux and I made during our testing session:

On the notes toggle button:

* ```<nav class="nav-utilities">``` needs to have an aria-label on it.  I suggest ```aria-label=Course Utilities```  You can then remove the SR only ```<h2>``` that follows.
* Nit: The status message implies that something with an indefinite end is happening "Hiding Notes"".  We think it should describe what just happened i.e. Notes On/Notes Off or Notes Enabled/Notes Disabled.
* ```aria-pressed```  is not necessary on the button since the state and action are (more) clear from the accessible text
* Since we need to make that Notes toggle as easy to find as possible and the keyboard shortcut might not work out, I suggest making the first word in the accessible label consistent.  Right now, it changes from Hiding notes to Showing notes.  The screen reader user will try to navigate directly to the button and has to remember what its currently called to do so.  Giving it a consistent naming convention like "Notes - Enable" and 
Notes - Disabled" allows them to always look for the button that starts with "Notes"

On the annotated text:

* When using a screen reader, just making the text focusable with ```tabindex=0``` and adding keyboard listeners is not enough.  We need to add a ```role="link"``` to it.  without that, keyboard focus is not moved to the text, only browser focus.
* When the note/annotation window pops open, we need to move keyboard focus to it.  I suggest moving focus to the container ```<div>``` which will require a ```tabindex="-1"```
* When focus is moved to this ```<div>``` a keyboard user should not be able to tab out of it without explicitly exiting with either a button press (a close button does not currently exist) or pressing ```ESC``` essentially trapping the keyboard focus.  Here is some more guidance on [creating accessible modals](https://github.com/edx/edx-documentation/blob/carol/docs/accessibility/en_us/developers/source/accessibility.rst#manage-the-focus-for-pop-ups)
* The icon for deleting a note is an ```X``` which is often associated with closing a dialog box.  We think this should be a trashcan.  We also think an ```X``` that closes the annotation should be added.
* Ideally, the icons would follow the text in the DOM order, but that is just a NIT. Right now, when a user encounters the note, they access the edit, delete and close buttons *before* the actual note.
* When you Edit a note, focus needs to be moved to this dialog as well, just like the note dialog itself (we noticed these are actually two separate modals)
* The ```<textarea>``` for editing a note does not have an accessible label.  Can we add ```<label class="sr">Edit note<label>``` to create a hidden accessible label?  This would require a ```for``` attribute on the label and a unique ```id``` on the ```<textarea>``` so this would also work: ```<label><span class="sr">Edit note</span><textarea></textarea><label>``` and wouldn't require the the ```id``` and ```for```

@explorerleslie

polesye commented:
> On the notes toggle button:
 - Nit: The status message implies that something with an indefinite end is happening "Hiding Notes"". We think it should describe what just happened i.e. Notes On/Notes Off or Notes Enabled/Notes Disabled.
 - Since we need to make that Notes toggle as easy to find as possible and the keyboard shortcut might not work out, I suggest making the first word in the accessible label consistent. Right now, it changes from Hiding notes to Showing notes. The screen reader user will try to navigate directly to the button and has to remember what its currently called to do so. Giving it a consistent naming convention like "Notes - Enable" and Notes - Disabled" allows them to always look for the button that starts with "Notes"

@srpearce  What do you think about the changes above?

jmclaus commented:
@cptvitamin @clrux Thanks for these very thorough comments. I've addressed part of them:

>* When using a screen reader, just making the text focusable with ```tabindex=0``` and adding keyboard listeners is not enough.  We need to add a ```role="link"``` to it.  without that, keyboard focus is not moved to the text, only browser focus.

Done.

> * When the note/annotation window pops open, we need to move keyboard focus to it.  I suggest moving focus to the container ```<div>``` which will require a ```tabindex="-1"```

Done. Except I can't manage to give focus to that ```div``` programmatically, even after adding the  ```tabindex="-1"``` attribute to it. So for the moment, when the viewer popup shows, I give focus to the ```edit``` button.

>* When focus is moved to this ```<div>``` a keyboard user should not be able to tab out of it without explicitly exiting with either a button press (a close button does not currently exist) or pressing ```ESC``` essentially trapping the keyboard focus.  Here is some more guidance on [creating accessible modals](https://github.com/edx/edx-documentation/blob/carol/docs/accessibility/en_us/developers/source/accessibility.rst#manage-the-focus-for-pop-ups)

Opening the viewer will give focus on the ```edit``` button and the tabbing will effectively be trapped, cycling through the ```edit```, ```delete```, and ```close``` buttons.

>* The icon for deleting a note is an ```X``` which is often associated with closing a dialog box.  We think this should be a trashcan.  We also think an ```X``` that closes the annotation should be added.

A very good idea, the ```X``` was confusing. I added a close button represented by a x. So, the following buttons now appear from left to right: ```edit``` (pencil), ```delete``` (trashcan), ```close``` (x).

@polesye FYI.

jmclaus commented:
Closing this PR in favor of https://github.com/edx/edx-platform/pull/6918

cptvitamin commented:
These are very exciting improvements @jmclaus.  Thanks for your work.  RE: focus.  Are you sure that focus was not being moved to the div?  The idea is that the next press of the tab key will move focus to the first focusable element inside.  I like that approach (moving focus to the thing just *before* the thing you really want focus to be on) because screen readers do *not* announce the element they land on when focus is programmatically moved.  A screen reader and keyboard only user will initiate the Note, then press tab and focus will be moved to the textarea and it will be announced.

jmclaus commented:
@cptvitamin For some reason, I cannot programmatically move the focus to the the viewer div. So for the moment, when the note opens, I give focus to the text area div. So the tabbing will cycle as following: Press ENTER/SPACE on highlight text. That gives focus to the textarea part of the viewer. Tabbing then cycles forwards or backwards through the following elements Text area/Edit/Delete/Close . So you are telling me that this wouldn't work with screen readers? @polesye FYI.

cptvitamin commented:
@jmclaus I was just confused as to why you couldn't programmatically move focus to the viewer div.  Were you getting an error?  If the div has a tabindex="-1" attribute, you can just call focus() on it.  Then, pressing the tab key should move focus to the textarea next.

jmclaus commented:
@cptvitamin That's exactly what I was trying to do and it doesn't work (tabindex = -1 has been correctly set on the div) 

https://github.com/edx/edx-platform/pull/6835/files#diff-16413263a60d575b2763fdf08127e484R148

jmclaus commented:
@cptvitamin But I'll try again.

cptvitamin commented:
@jmclaus where does Inspector show focus is?  I guess I'm wondering how you conclude that focus is not being moved.  Could the subsequent Tab press not be working?  Could that be intercepted by this line: https://github.com/edx/edx-platform/pull/6835/files#diff-16413263a60d575b2763fdf08127e484R17

jmclaus commented:
@cptvitamin The viewer could not get the focus because for some reason, the creators of AnnotatorJS give it a size of 0px x 0px. But it is possible to focus on the next element below it in the DOM tree, ul.annotator-listing if we give it a tab index = -1. Which I just implemented. So the tabbing will now cycle as following: Press ENTER/SPACE on highlight text. That gives focus to ul.annotator-listing. Tabbing forward gives focus to textarea. Tabbing backwards gives focus to close. Once focused on one of these elements, tabbing then cycles forwards or backwards through the following elements: Text area/Edit/Delete/Close.

@polesye FYI.

jmclaus commented:
@polesye ye @cptvitamin @clrux I have cherry-picked all the commits from this branch and have applied them to the tip of a freshly rebased jmclaus/edxnotes-add-a11y. The associated PR is: https://github.com/edx/edx-platform/pull/6778 Lets continue discussion and review there. We will now have 2 PRs that deals with notes' a11y:

* On notes already created: https://github.com/edx/edx-platform/pull/6778
* With creation of notes: https://github.com/edx/edx-platform/pull/6744

cptvitamin commented:
Thanks for all your work on this @jmclaus.  Thanks for working through that focus and tab order issue.  Great work!

jmclaus commented:
@clrux I fixed the Jasmine tests after yet another change in keyboard interaction. Tests pass locally. So I think you'll be good to merge your branch into this one once you've rebased. Thanks.

jmclaus commented:
@clrux See https://github.com/edx/edx-platform/pull/6778
